src/tweenable.rs

```diff
@@ -894,264 +917,399 @@ mod tests {
+
+                let mut event_reader = event_reader.get_mut(&mut world);
+                if just_completed {
+                    assert!(!event_reader.is_empty());
+                    for event in event_reader.iter() {
+                        assert_eq!(dummy_entity, event.entity);
+                        assert_eq!(USER_DATA, event.user_data, "user_data");
+                    }
+                } else {
+                    assert!(event_reader.is_empty());
+                }
             }
         }
     }
 
+    #[rstest]
+    fn tween_tick_with_defaults_and_directions(
+        #[values(TweeningDirection::Forward, TweeningDirection::Backward)]
+        direction: TweeningDirection,
+    ) {
+        let tween = Tween::new(
+            EaseMethod::Linear,
+            Duration::from_secs(1),
+            TransformPositionLens {
+                start: Vec3::ZERO,
+                end: Vec3::ONE,
+            },
+        )
+        .with_direction(direction);
+
+        let expected_values = ExpectedValues {
+            deltas: once(Duration::ZERO).chain(repeat(Duration::from_millis(200)).take(6)),
+            progress: successors(Some(0.), |progress| Some(f32::min(progress + 0.2, 1.))),
+            times_completed: [0, 0, 0, 0, 0, 1, 1],
+            event_counts: [0, 0, 0, 0, 0, 1, 1],
+            directions: repeat(direction),
+            states: repeat(TweenState::Active)
+                .take(5)
+                .chain(repeat(TweenState::Completed)),
+            transforms: if direction == TweeningDirection::Forward {
+                successors(Some(0.), |progress| Some(f32::min(progress + 0.2, 1.)))
+                    .map(|progress| Transform::from_translation(Vec3::splat(progress)))
+                    .take(7)
+                    .collect::<Vec<_>>()
+            } else {
+                successors(Some(1.), |progress| Some(f32::max(progress - 0.2, 0.)))
+                    .map(|progress| Transform::from_translation(Vec3::splat(progress)))
+                    .take(7)
+                    .collect::<Vec<_>>()
+            },
+        };
+
+        validate_tween(tween, expected_values);
+    }
+
+    #[test]
+    fn tween_tick_loop_finite() {
+        let tween = Tween::new(
+            EaseMethod::Linear,
+            Duration::from_secs(1),
+            TransformPositionLens {
+                start: Vec3::ZERO,
+                end: Vec3::ONE,
+            },
+        )
+        .with_direction(TweeningDirection::Forward)
+        .with_repeat_count(RepeatCount::Finite(3))
+        .with_repeat_strategy(RepeatStrategy::Repeat);
+
+        let expected_values = ExpectedValues {
+            deltas: once(Duration::ZERO).chain(repeat(Duration::from_secs_f32(1. / 3.)).take(10)),
+            progress: successors(Some(0.), |progress| Some(f32::min(3., progress + 1. / 3.))),
+            // TODO this is totally wrong due to float precision, should be fixed in upcoming
```

Comment:
How do you want to address that? Can we just avoid the edge case by changing the deltas? I understand that 3 steps of 1/3 second might or might not complete the first 1-second loop, but that's just how floating point arithmetic works, we can't really make any strong guarantee for that in the API, can we?

Reply:
Like, if we do `delta = 0.32` then we should avoid any floating-point error over 10+ loops, so would have a deterministic test.

Reply:
Well in theory this should work because we're using durations directly, so I'm hoping #38 will fix it. If not, I'll investigate what's going on and we can probably just use 0.32 like you said.

Reply:
Oh wait no I see what you mean, the `1. / 3.` is what's hosing us. Will tweak it a little bit.